File: solutions/3100. Water Bottles II/3100.py
# Time complexity: O(n)
# Space complexity: O(1)
class Solution:
  def maxBottlesDrunk(self, numBottles: int, numExchange: int) -> int:
    count = numBottles

    while numBottles >= numExchange:
      numBottles -= numExchange - 1
      numExchange += 1
      count += 1

    return count

File: solutions/3100. Water Bottles II/3100.cpp
// Time complexity: O(n)
// Space complexity: O(1)
class Solution {
 public:
  int maxBottlesDrunk(int numBottles, int numExchange) {
    int count = numBottles;

    while (numBottles >= numExchange) {
      numBottles -= numExchange - 1;
      ++numExchange;
      ++count;
    }

    return count;
  }
};